jsprit-core/src/main/java/com/graphhopper/jsprit/core/problem/job/AbstractJob.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to GraphHopper GmbH under one or more contributor
+ * license agreements. See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ *
+ * GraphHopper GmbH licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.graphhopper.jsprit.core.problem.job;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import com.graphhopper.jsprit.core.problem.Location;
+import com.graphhopper.jsprit.core.problem.solution.route.activity.TimeWindow;
+
+/**
+ * Created by schroeder on 14.07.14.
+ */
+public abstract class AbstractJob implements Job {
+
+    private int index;
+
+    protected List<Location> allLocations = new ArrayList<>();
+
+    private JobActivityList activityList;
+
+    protected Set<TimeWindow> operationTimeWindows = new HashSet<>();
+
+    public AbstractJob() {
+        super();
+        activityList = new SequentialJobActivityList(this);
+    }
+
+    @Override
+    public int getIndex() {
+        return index;
+    }
+
+    public void setIndex(int index) {
+        this.index = index;
+    }
+
+    protected void addLocation(Location location) {
```

Comment:
**No, we don't need `addLocations` and `addTimeWindows`.** Yesterday (oh, it was today midnight) I've finally understood your first comment and you were absolutely right.
All this information could be automatically collected from the created activities, so it must be do in that way. 
In that way, a new Job implementation only need to override the sole `createActivities()` method. 
For the TW and location registration, there are two possible ways:

- Make a contract in the javadoc of createActivities to call a single function (such as `prepareCaches()`)
- Or make the cache initialization on-demand (when the getters are called at the first time.

The first adds an additional task for the user, the second has an overhead, because every time a cache getter is called it has to be checked whether the cache is initialized. I vote for the first one, but I would like to hear your opinion.